src/statsdproto.rs

```diff
@@ -0,0 +1,385 @@
+use bytes::BufMut;
+use bytes::Bytes;
+use memchr::{memchr, memchr2};
+use thiserror::Error;
+
+use std::{
+    convert::{TryFrom, TryInto},
+    vec,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Type {
+    Counter,
+    Timer,
+    Gauge,
+    DirectGauge,
+    Set,
+}
+
+impl TryFrom<&[u8]> for Type {
+    type Error = ParseError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() < 1 && value.len() > 2 {
+            return Err(ParseError::InvalidType);
+        }
+        match value[0] {
+            b'c' => Ok(Type::Counter),
+            b'm' => Ok(Type::Timer),
+            b'g' => Ok(Type::Gauge),
+            b'G' => Ok(Type::DirectGauge),
+            b's' => Ok(Type::Set),
+            _ => Err(ParseError::InvalidType),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("invalid parsed value")]
+    InvalidValue,
+    #[error("invalid sample rate")]
+    InvalidSampleRate,
+    #[error("invalid type")]
+    InvalidType,
+    #[error("invalid tag")]
+    InvalidTag,
+}
+
+/// A Tag is a key:value pair of metric tags
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+pub struct Tag {
+    name: Vec<u8>,
+    value: Vec<u8>,
+}
+
+/// Parsed gives an owned structure which represents a parsed statsd protocol
+/// unit which owns all of its fields. When parsing, no canonicalization is
+/// performed by default.
```

Comment:
Another minor docs nit: IME with types that're basically only meant to be used via a trait method, especially a stdlib trait, it's super helpful to have a `usage:` section in the docstring and a rustdoc link to the type or function they're coupled to. That's especially helpful for folks who're new to rust/rustdoc since they won't necessarily be super comfortable scanning the trait list to find something that sticks out.

e.g. This is the doc output for `Parsed` as of this commit:
![image](https://user-images.githubusercontent.com/596042/113919833-44849300-97a1-11eb-820b-7c3603122a29.png)